File: bindings/python/cstest_py/src/cstest_py/compare.py
# Copyright © 2024 Rot127 <unisono@quyllur.org>
# SPDX-License-Identifier: BSD-3

# Typing for Python3.8
from __future__ import annotations

import struct
import capstone
import re
from capstone import arm_const
from capstone import aarch64_const
from capstone import m68k_const
from capstone import mips_const
from capstone import ppc_const
from capstone import sparc_const
from capstone import systemz_const
from capstone import x86_const
from capstone import xcore_const
from capstone import tms320c64x_const
from capstone import m680x_const
from capstone import evm_const
from capstone import mos65xx_const
from capstone import wasm_const
from capstone import bpf_const
from capstone import riscv_const
from capstone import sh_const
from capstone import tricore_const
from capstone import alpha_const
from capstone import hppa_const
from capstone import loongarch_const
from capstone import arc_const


def cs_const_getattr(identifier: str):
    attr = getattr(capstone, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(arm_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(aarch64_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(m68k_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(mips_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(ppc_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(sparc_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(systemz_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(x86_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(xcore_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(tms320c64x_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(m680x_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(evm_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(mos65xx_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(wasm_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(bpf_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(riscv_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(sh_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(tricore_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(alpha_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(hppa_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(loongarch_const, identifier, None)
    if attr is not None:
        return attr
    attr = getattr(arc_const, identifier, None)
    if attr is not None:
        return attr
    raise ValueError(f"Python capstone doesn't have the constant: {identifier}")


def twos_complement(val, bits):
    if (val & (1 << (bits - 1))) != 0:
        val = val - (1 << bits)
    return val & ((1 << bits) - 1)


def normalize_asm_text(text: str, arch_bits: int) -> str:
    text = text.strip()
    text = re.sub(r"\s+", " ", text)
    # Replace hex numbers with decimals
    for hex_num in re.findall(r"0x[0-9a-fA-F]+", text):
        text = re.sub(hex_num, f"{int(hex_num, base=16)}", text, count=1)
    # Replace negatives with twos-complement
    for num in re.findall(r"-\d+", text):
        n = twos_complement(int(num, base=10), arch_bits)
        text = re.sub(num, f"{n}", text)
    text = text.lower()
    return text


def compare_asm_text(
    a_insn: capstone.CsInsn, expected: None | str, arch_bits: int
) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    actual = f"{a_insn.mnemonic} {a_insn.op_str}"
    actual = normalize_asm_text(actual, arch_bits)
    expected = normalize_asm_text(expected, arch_bits)

    if actual != expected:
        log.error(
            "Normalized asm-text doesn't match:\n"
            f"decoded:  '{actual}'\n"
            f"expected: '{expected}'\n"
        )
        return False
    return True


def compare_str(actual: str, expected: None | str, msg: str) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    if actual != expected:
        log.error(f"{msg}: {actual} != {expected}")
        return False
    return True


def compare_tbool(actual: bool, expected: None | int, msg: str) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    if expected == 0:
        # Unset
        return True

    if (expected < 0 and actual) or (expected > 0 and not actual):
        log.error(f"{msg}: {actual} != {expected}")
        return False
    return True


def compare_uint8(actual: int, expected: None | int, msg: str) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    actual = actual & 0xFF
    expected = expected & 0xFF
    if actual != expected:
        log.error(f"{msg}: {actual} != {expected}")
        return False
    return True


def compare_int8(actual: int, expected: None | int, msg: str) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    actual = actual & 0xFF
    expected = expected & 0xFF
    if actual != expected:
        log.error(f"{msg}: {actual} != {expected}")
        return False
    return True


def compare_uint16(actual: int, expected: None | int, msg: str) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    actual = actual & 0xFFFF
    expected = expected & 0xFFFF
    if actual != expected:
        log.error(f"{msg}: {actual} != {expected}")
        return False
    return True


def compare_int16(actual: int, expected: None | int, msg: str) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    actual = actual & 0xFFFF
    expected = expected & 0xFFFF
    if actual != expected:
        log.error(f"{msg}: {actual} != {expected}")
        return False
    return True


def compare_uint32(actual: int, expected: None | int, msg: str) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    actual = actual & 0xFFFFFFFF
    expected = expected & 0xFFFFFFFF
    if actual != expected:
        log.error(f"{msg}: {actual} != {expected}")
        return False
    return True


def compare_int32(actual: int, expected: None | int, msg: str) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    actual = actual & 0xFFFFFFFF
    expected = expected & 0xFFFFFFFF
    if actual != expected:
        log.error(f"{msg}: {actual} != {expected}")
        return False
    return True


def compare_uint64(actual: int, expected: None | int, msg: str) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    actual = actual & 0xFFFFFFFFFFFFFFFF
    expected = expected & 0xFFFFFFFFFFFFFFFF
    if actual != expected:
        log.error(f"{msg}: {actual} != {expected}")
        return False
    return True


def compare_int64(actual: int, expected: None | int, msg: str) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    actual = actual & 0xFFFFFFFFFFFFFFFF
    expected = expected & 0xFFFFFFFFFFFFFFFF
    if actual != expected:
        log.error(f"{msg}: {actual} != {expected}")
        return False
    return True


def compare_fp(actual: float, expected: None | float, msg: str) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    def floatToBits(f):
        return struct.unpack("=L", struct.pack("=f", f))[0]

    if floatToBits(actual) != floatToBits(expected):
        log.error(f"{msg}: {actual} != {expected}")
        return False
    return True


def compare_dp(actual: float, expected: None | float, msg: str) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    def doubleToBits(f):
        return struct.unpack("=Q", struct.pack("=d", f))[0]

    if doubleToBits(actual) != doubleToBits(expected):
        log.error(f"{msg}: {actual} != {expected}")
        return False
    return True


def compare_enum(actual, expected: None | str, msg: str) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    enum_val = cs_const_getattr(expected)
    if actual != enum_val:
        log.error(f"{msg}: {actual} != {expected} ({enum_val})")
        return False
    return True


def compare_bit_flags(actual: int, expected: None | list[str], msg: str) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    for flag in expected:
        enum_val = cs_const_getattr(flag)
        if not actual & enum_val:
            log.error(f"{msg}: In {actual:x} the flag {expected} isn't set.")
            return False
    return True


def compare_reg(
    insn: capstone.CsInsn, actual: int, expected: None | str, msg: str
) -> bool:
    if expected is None:
        return True
    from cstest_py.cstest import log

    if insn.reg_name(actual) != expected:
        log.error(f"{msg}: {actual} != {expected}")
        return False
    return True

File: bindings/java/cstest_java/app/src/main/java/org/capstone_engine/cstest/Compare.java
// Copyright © 2025 Peace-Maker <peacemakerctf@gmail.com>
// SPDX-License-Identifier: BSD-3
package org.capstone_engine.cstest;

import java.math.BigInteger;
import java.util.List;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import capstone.Capstone;
import capstone.Arm_const;
import capstone.AArch64_const;
import capstone.M68k_const;
import capstone.Mips_const;
import capstone.Ppc_const;
import capstone.Sparc_const;
import capstone.Systemz_const;
import capstone.X86_const;
import capstone.Xcore_const;
import capstone.TMS320C64x_const;
import capstone.M680x_const;
import capstone.Evm_const;
import capstone.Mos65xx_const;
import capstone.Wasm_const;
import capstone.Bpf_const;
import capstone.Riscv_const;
import capstone.Sh_const;
import capstone.Tricore_const;
import capstone.Alpha_const;
import capstone.Hppa_const;
import capstone.Loongarch_const;
import capstone.Xtensa_const;
import capstone.Arc_const;


public class Compare {
    private static final Logger log = Logger.getLogger(Compare.class.getName());

    private static Long getLongAttr(Class<?> klass, String attr) {
        try {
            return klass.getField(attr).getLong(null);
        } catch (NoSuchFieldException | IllegalAccessException e) {
            return null;
        }
    }

    public static long cs_const_getattr(String identifier) {
        Long attr = getLongAttr(Capstone.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Arm_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(AArch64_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(M68k_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Mips_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Ppc_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Sparc_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Systemz_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(X86_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Xcore_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(TMS320C64x_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(M680x_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Evm_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Mos65xx_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Wasm_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Bpf_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Riscv_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Sh_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Tricore_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Alpha_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Hppa_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Loongarch_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Xtensa_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        attr = getLongAttr(Arc_const.class, identifier);
        if (attr != null) {
            return attr;
        }
        throw new IllegalArgumentException("Java capstone doesn't have the constant: " + identifier);
    }

    public static long twosComplement(long val, long bits) {
        if ((val & (1L << (bits - 1L))) != 0L) {
            val = val - (1L << bits);
        }
        return val & ((1L << bits) - 1L);
    }

    public static String normalizeAsmText(String text, int archBits) {
        text = text.trim();
        text = text.replaceAll("\\s+", " ");

        // Replace hex numbers with decimals
        Pattern hexPattern = Pattern.compile("0x[0-9a-fA-F]+");
        Matcher hexMatcher = hexPattern.matcher(text);
        while (hexMatcher.find()) {
            String hexNum = hexMatcher.group();
            long decimalValue = new BigInteger(hexNum.substring(2), 16).longValue();
            text = text.replaceFirst(hexNum, String.valueOf(decimalValue));
        }

        // Replace negatives with two's complement
        Pattern numPattern = Pattern.compile("-\\d+");
        Matcher numMatcher = numPattern.matcher(text);
        while (numMatcher.find()) {
            String num = numMatcher.group();
            long n = twosComplement(Long.parseLong(num), archBits);
            text = text.replaceFirst(Pattern.quote(num), String.valueOf(n));
        }

        text = text.toLowerCase();
        return text;
    }

    public static boolean compareAsmText(Capstone.CsInsn aInsn, String expected, int archBits) {
        if (expected == null) {
            return true;
        }

        String actual = aInsn.mnemonic + " " + aInsn.opStr;
        actual = normalizeAsmText(actual, archBits);
        expected = normalizeAsmText(expected, archBits);

        if (!actual.equals(expected)) {
            log.severe("Normalized asm-text doesn't match:\n" +
                    "decoded:  '" + actual + "'\n" +
                    "expected: '" + expected + "'");
            return false;
        }
        return true;
    }
    
    public static boolean compareStr(String actual, String expected, String msg) {
        if (expected == null) {
            return true;
        }

        if (!actual.equals(expected)) {
            log.severe(msg + ": " + actual + " != " + expected);
            return false;
        }
        return true;
    }

    public static boolean compareTBool(boolean actual, Integer expected, String msg) {
        if (expected == null) {
            return true;
        }

        if (expected == 0) {
            // Unset
            return true;
        }

        if ((expected < 0 && actual) || (expected > 0 && !actual)) {
            log.severe(msg + ": " + actual + " != " + expected);
            return false;
        }
        return true;
    }

    public static boolean compareUInt8(int actual, Integer expected, String msg) {
        if (expected == null) {
            return true;
        }

        actual = actual & 0xFF;
        expected = expected & 0xFF;
        if (actual != expected) {
            log.severe(msg + ": " + actual + " != " + expected);
            return false;
        }
        return true;
    }

    public static boolean compareInt8(int actual, Integer expected, String msg) {
        if (expected == null) {
            return true;
        }

        actual = actual & 0xFF;
        expected = expected & 0xFF;
        if (actual != expected) {
            log.severe(msg + ": " + actual + " != " + expected);
            return false;
        }
        return true;
    }

    public static boolean compareUInt16(int actual, Integer expected, String msg) {
        if (expected == null) {
            return true;
        }

        actual = actual & 0xFFFF;
        expected = expected & 0xFFFF;
        if (actual != expected) {
            log.severe(msg + ": " + actual + " != " + expected);
            return false;
        }
        return true;
    }

    public static boolean compareInt16(int actual, Integer expected, String msg) {
        if (expected == null) {
            return true;
        }

        actual = actual & 0xFFFF;
        expected = expected & 0xFFFF;
        if (actual != expected) {
            log.severe(msg + ": " + actual + " != " + expected);
            return false;
        }
        return true;
    }

    public static boolean compareUInt32(int actual, Integer expected, String msg) {
        if (expected == null) {
            return true;
        }

        actual = actual & 0xFFFFFFFF;
        expected = expected & 0xFFFFFFFF;
        if (actual != expected) {
            log.severe(msg + ": " + actual + " != " + expected);
            return false;
        }
        return true;
    }

    public static boolean compareInt32(int actual, Integer expected, String msg) {
        if (expected == null) {
            return true;
        }

        actual = actual & 0xFFFFFFFF;
        expected = expected & 0xFFFFFFFF;
        if (actual != expected) {
            log.severe(msg + ": " + actual + " != " + expected);
            return false;
        }
        return true;
    }

    public static boolean compareUInt64(long actual, Long expected, String msg) {
        if (expected == null) {
            return true;
        }

        actual = actual & 0xFFFFFFFFFFFFFFFFL;
        expected = expected & 0xFFFFFFFFFFFFFFFFL;
        if (actual != expected) {
            log.severe(msg + ": " + actual + " != " + expected);
            return false;
        }
        return true;
    }

    public static boolean compareInt64(long actual, Long expected, String msg) {
        if (expected == null) {
            return true;
        }

        actual = actual & 0xFFFFFFFFFFFFFFFFL;
        expected = expected & 0xFFFFFFFFFFFFFFFFL;
        if (actual != expected) {
            log.severe(msg + ": " + actual + " != " + expected);
            return false;
        }
        return true;
    }

    public static boolean compareFp(float actual, Float expected, String msg) {
        if (expected == null) {
            return true;
        }

        if (Float.floatToIntBits(actual) != Float.floatToIntBits(expected)) {
            log.severe(msg + ": " + actual + " != " + expected);
            return false;
        }
        return true;
    }

    public static boolean compareDp(double actual, Double expected, String msg) {
        if (expected == null) {
            return true;
        }

        if (Double.doubleToLongBits(actual) != Double.doubleToLongBits(expected)) {
            log.severe(msg + ": " + actual + " != " + expected);
            return false;
        }
        return true;
    }

    public static boolean compareEnum(long actual, String expected, String msg) {
        if (expected == null) {
            return true;
        }

        long enumVal = cs_const_getattr(expected);
        if (actual != enumVal) {
            log.severe(msg + ": " + actual + " != " + expected + " (" + enumVal + ")");
            return false;
        }
        return true;
    }

    public static boolean compareBitFlags(long actual, List<String> expected, String msg) {
        if (expected == null) {
            return true;
        }

        for (String flag : expected) {
            long enumVal = cs_const_getattr(flag);
            if ((actual & enumVal) == 0L) {
                log.severe(msg + ": In " + Long.toHexString(actual) + " the flag " + flag + " isn't set.");
                return false;
            }
        }
        return true;
    }

    public static boolean compareReg(Capstone.CsInsn insn, int actual, String expected, String msg) {
        if (expected == null) {
            return true;
        }

        if (insn.regName(actual) == null) {
            log.severe(msg + ": " + actual + " isn't a valid register id");
            return false;
        }

        if (!insn.regName(actual).equals(expected)) {
            log.severe(msg + ": " + actual + " != " + expected);
            return false;
        }
        return true;
    }
}
